crates/codegen/runtime/cargo/src/runtime/parse_error.rs

```diff
@@ -43,52 +39,41 @@ impl ParseError {
     }
 }
 
-pub(crate) fn render_error_report(
-    error: &ParseError,
-    source_id: &str,
-    source: &str,
-    with_color: bool,
-) -> String {
-    use ariadne::{Color, Config, Label, Report, ReportKind, Source};
-
-    let kind = ReportKind::Error;
-    let color = if with_color { Color::Red } else { Color::Unset };
-
-    let tokens_that_would_have_allowed_more_progress =
-        error.tokens_that_would_have_allowed_more_progress();
-    let message = if tokens_that_would_have_allowed_more_progress.is_empty() {
-        "Expected end of file.".to_string()
-    } else {
-        format!(
-            "Expected {expectations}.",
-            expectations = tokens_that_would_have_allowed_more_progress.join(" or ")
-        )
-    };
-
-    if source.is_empty() {
-        return format!("{kind}: {message}\n   ─[{source_id}:0:0]");
+impl Error for ParseError {}
+impl fmt::Display for ParseError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if self.tokens_that_would_have_allowed_more_progress.is_empty() {
+            write!(f, "Expected end of file.")
+        } else {
+            let deduped = self
+                .tokens_that_would_have_allowed_more_progress
```

Comment:
Unrelated to this PR, but TOL: I wonder if we should deduplicate during construction by using a set in `ParseError`, if this is a common scenario.

Reply:
We should and I had this at the back of my head for quite a bit but that's unrelated, so I'd prefer not to mix it in here.